File: cloneGraph.cpp
// https://leetcode.com/problems/clone-graph/

#include <bits/stdc++.h>
using namespace std;

class Node {
 public:
  int val;
  vector<Node*> neighbors;
  Node() {
    val = 0;
    neighbors = vector<Node*>();
  }
  Node(int _val) {
    val = _val;
    neighbors = vector<Node*>();
  }
  Node(int _val, vector<Node*> _neighbors) {
    val = _val;
    neighbors = _neighbors;
  }
};

class Solution {
 public:
  unordered_map<int, Node*> visited;
  Node* cloneGraph(Node* node) {
    if(!node){
      return node;
    }
    if (visited.find(node->val) == visited.end()) {
      Node* newNode = new Node(node->val);
      visited.insert({newNode->val, newNode});
      for (int i = 0; i < node->neighbors.size(); i++) {
        newNode->neighbors.push_back(cloneGraph(node->neighbors[i]));
      }
      return newNode;
    } else {
      return visited.find(node->val)->second;
    }
  }
};

File: grind75_python/cloneGraph.py
# https://leetcode.com/problems/clone-graph/description/

# Definition for a Node.
from typing import Deque, Dict, Optional
from collections import deque


class Node:
  def __init__(self, val=0, neighbors=None):
    self.val = val
    self.neighbors = neighbors if neighbors is not None else []


class Solution:
  def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
    if not node:
      return node

    created: Dict[int, Node] = dict()
    q: Deque[Node] = deque()

    q.append(node)
    ans = Node(node.val)
    created[1] = ans

    while len(q) != 0:
      curr = q.popleft()
      for n in curr.neighbors:
        if not n.val in created:
          new_node = Node(n.val)
          created[new_node.val] = new_node
          q.append(n)
        created[curr.val].neighbors.append(created[n.val])

    return ans
